### ISSUE TITLE
Performance regressions on v3

### ISSUE THREAD

**Issue by radekmie** [vazco/uniforms#716]
Step by step, commit by commit - we're getting closer to a state, where v3 could be released (finally!). You don't have to run benchmarks to see, that the new version is far slower! Let's dissect the problems and think about possible solutions.

### Consequences of new context API
Using `useField`, or `useForm` directly, relies on `useContext` underneath. While the new context API is far better and more composable, it differs in the way of context propagation. The old context was more or less another set of props. This caused a lot of issues with other packages using `PureComponent`/`memo` (like #528). New one updates all consumers directly, skipping all nodes in between. This means, that there's no way to actually stop an update on the consumer side, and existing [`shouldComponentUpdate`](https://github.com/vazco/uniforms/blob/04e1ff427998f1369f39945ad416f443c2c0ff7a/packages/uniforms/src/BaseField.tsx#L93-L169) won't help us now.

As there's no React way to deal with that _yet_ (there are RFCs though, like https://github.com/reactjs/rfcs/pull/119), we have to do _something_. There are userspace implementations, like [dai-shi/use-context-selector](https://github.com/dai-shi/use-context-selector). I think we could use that, but the actual gains need to be confirmed in benchmarks. This would also mean, that the actual context usage may change with time, and we should no longer treat the context as a part of public API (of course, `useForm` and `useField` are still available).

### Consequences of new lifecycle APIs
Using `componentWillReceiveProps` one can access both previous and next props as well as the current state. New `getDerivedStateFromProps` differs a little: there's no access to previous props. This means, that the `BaseForm` cannot recreate the bridge only when needed! [I've signalized it already](https://github.com/vazco/uniforms/blob/0ad207a677e1566d426d0b60a206360808a46f80/packages/uniforms/src/BaseForm.tsx#L86-L92) and we should consider actually implementing that.

**EDIT:** Other solution can be actually removing `createSchemaBridge` completely. This would eliminate the awkward `Bridge.check` as well as the bridge duplication problem because one would have to create the bridge manually beforehand. It leads to a slightly more complicated usage with SimpleSchema, but it changes nothing with JSON schema GraphQL and schemas.

### Other problems
I'm sure there are more things we could do. We have to take the opportunity of making breaking changes (as small as possible!) and look for them now. One thing we could take a look at is the `modelSync` of `AutoForm` - it may be not needed with the Suspend API.

**Comment by radekmie** [comment]
Changes made in e27ea076c7b631c7e498d15d5750bf8da945dbf3, ddba4da3e35880c4deb1f2c47bd48a7e8553a65c, and 69d2b78784d1fcf49793af96e290f4e92c97259a improved performance a lot for big schemas.

|  | Before | After |
| - | -:| -:|
| min | 78ms | 42ms |
| max | 116ms | 70ms |

<details>
<summary>Tested schema.</summary>

```tsx
new SimpleSchema2Bridge(
  new SimpleSchema({
    a: String,
    b: String,
    c: String,
    d: String,
    e: new SimpleSchema({
      a: String,
      b: String,
      c: String,
      d: String,
      e: new SimpleSchema({
        a: String,
        b: String,
        c: String,
        d: String,
        e: new SimpleSchema({
          a: String,
          b: String,
          c: String,
          d: String,
          e: new SimpleSchema({
            a: String,
            b: String,
            c: String,
            d: String,
            e: new SimpleSchema({
              a: String,
              b: String,
              c: String,
              d: String,
              e: new SimpleSchema({
                a: String,
                b: String,
                c: String,
                d: String,
                e: new SimpleSchema({
                  a: String,
                  b: String,
                  c: String,
                  d: String,
                  e: new SimpleSchema({
                    a: String,
                    b: String,
                    c: String,
                    d: String,
                    e: new SimpleSchema({
                      a: String,
                      b: String,
                      c: String,
                      d: String,
                      e: new SimpleSchema({
                        a: String,
                        b: String,
                        c: String,
                        d: String,
                        e: new SimpleSchema({
                          a: String,
                          b: String,
                          c: String,
                          d: String,
                          e: new SimpleSchema({
                            a: String,
                            b: String,
                            c: String,
                            d: String,
                            e: new SimpleSchema({
                              a: String,
                              b: String,
                              c: String,
                              d: String,
                              e: String
                            })
                          })
                        })
                      })
                    })
                  })
                })
              })
            })
          })
        })
      })
    })
  })
)
```
</details>

**Comment by radekmie** [comment]
With #739 and #717 (implemented in #741) the performance is already _good enough_. I've performed some experiments regarding the context and it's not helping us much as long as the schema is not very deep and narrow (which is usually not the case). There are two more things we should consider doing:
* Using [`memo`](https://reactjs.org/docs/react-api.html#reactmemo) when possible (especially for simple fields like `TextField`). It gives a big performance boost as the props rarely change. There's a problem though - `memo`ed components don't have explicit `name` nor `displayName`. There's an issue to have an API for that though: facebook/react#14319. Some quick experiments show massive improvements, especially with #741 in place.
* Use a custom `Provider` for the context. I wouldn't like to rework the way how we use context _now_ (maybe in v4...?), therefore each render produces a new context object, ultimately re-rendering all fields. What we could do is to make a custom wrapper for the provider that preserves the reference as long as the next one is equivalent (i.e. would render the same form). I've prepared some basic version and it can help a lot, especially when fields are deeply nested.

Both of these could be added later as they won't break the compatibility. Therefore, I consider this task done after #739 and #741 get merged.
